File: donors/orchestration-java/StateDiff.java
package com.aiassistant.core.orchestration;

import java.util.ArrayList;
import java.util.List;

public class StateDiff {
    private final String componentId;
    private final long timestamp;
    private final Severity severity;
    private final String description;
    private final List<FieldDiff> fieldDiffs;
    private final ComponentStateSnapshot expectedState;
    private final ComponentStateSnapshot actualState;
    
    public StateDiff(String componentId, Severity severity, String description,
                     List<FieldDiff> fieldDiffs, ComponentStateSnapshot expectedState,
                     ComponentStateSnapshot actualState) {
        this.componentId = componentId;
        this.severity = severity;
        this.description = description;
        this.fieldDiffs = fieldDiffs != null ? new ArrayList<>(fieldDiffs) : new ArrayList<>();
        this.expectedState = expectedState;
        this.actualState = actualState;
        this.timestamp = System.currentTimeMillis();
    }
    
    public String getComponentId() {
        return componentId;
    }
    
    public long getTimestamp() {
        return timestamp;
    }
    
    public Severity getSeverity() {
        return severity;
    }
    
    public String getDescription() {
        return description;
    }
    
    public List<FieldDiff> getFieldDiffs() {
        return new ArrayList<>(fieldDiffs);
    }
    
    public ComponentStateSnapshot getExpectedState() {
        return expectedState;
    }
    
    public ComponentStateSnapshot getActualState() {
        return actualState;
    }
    
    public enum Severity {
        INFO,
        WARNING,
        CRITICAL
    }
    
    public static class FieldDiff {
        public final String fieldName;
        public final Object expectedValue;
        public final Object actualValue;
        public final String diffType;
        
        public FieldDiff(String fieldName, Object expectedValue, Object actualValue, String diffType) {
            this.fieldName = fieldName;
            this.expectedValue = expectedValue;
            this.actualValue = actualValue;
            this.diffType = diffType;
        }
        
        @Override
        public String toString() {
            return "FieldDiff{" +
                    "field='" + fieldName + '\'' +
                    ", expected=" + expectedValue +
                    ", actual=" + actualValue +
                    ", type='" + diffType + '\'' +
                    '}';
        }
    }
    
    @Override
    public String toString() {
        return "StateDiff{" +
                "componentId='" + componentId + '\'' +
                ", severity=" + severity +
                ", description='" + description + '\'' +
                ", fieldDiffs=" + fieldDiffs.size() +
                '}';
    }
}

File: android/app/src/main/kotlin/com/ariaagent/mobile/core/orchestration/StateDiff.kt
// Ported from donors/orchestration-java/StateDiff.java
// Original repo: TITANICBHAI/AI-ASSISTANT-INCOMPLETE
// Changes: data class + nested data class for FieldDiff, enum class for Severity.

package com.ariaagent.mobile.core.orchestration

/**
 * The result of [DiffEngine.checkDiff] — describes the per-field deltas
 * between an expected and actual [ComponentStateSnapshot] plus a severity
 * tier the orchestrator can route on.
 */
data class StateDiff(
    val componentId: String,
    val severity: Severity,
    val description: String,
    val fieldDiffs: List<FieldDiff>,
    val expectedState: ComponentStateSnapshot,
    val actualState: ComponentStateSnapshot,
    val timestamp: Long = System.currentTimeMillis(),
) {
    enum class Severity { INFO, WARNING, CRITICAL }

    data class FieldDiff(
        val fieldName: String,
        val expectedValue: Any?,
        val actualValue: Any?,
        val diffType: DiffType,
    ) {
        enum class DiffType { VALUE_MISMATCH, UNEXPECTED_FIELD, MISSING_FIELD }
    }

    override fun toString(): String =
        "StateDiff(component=$componentId, severity=$severity, " +
            "description='$description', fields=${fieldDiffs.size})"
}
